src/mono/mono/mini/interp/transform.c

```diff
@@ -3320,6 +3320,88 @@ interp_try_devirt (MonoClass *this_klass, MonoMethod *target_method)
 	return NULL;
 }
 
+static MonoMethodSignature*
+interp_emit_swiftcall_struct_lowering (TransformData *td, MonoMethodSignature *csignature)
+{
+	MonoMethodSignature *new_csignature;
+	g_assert (!csignature->hasthis); // P/Invoke calls shouldn't contain 'this'
+
```

Comment:
The argument moving around that happens here will break the fast offset allocator (where we allocate storage for vars immediately as we push it on the `td->sp` stack). This appears however to be fine because we disable tiering for wrappers and I would imagine a swift call is only happening from a m2n wrapper. However I would assert here that `td->optimized` with the comment that argument reordering here doesn't handle on the fly offset allocation and requires the full var offset allocator pass that is only ran for optimized code.

Reply:
Yes, we only need to perform the struct lowering for m2n (and later eventually n2m) wrappers. I added the assert with the comment you suggested.